### PR TITLE
Feature dune v1.1 develop

### PR DESCRIPTION
**Pull request by EdAtkin**
Changing the DUNE code to be up to date with a new version of the core code. The version of the core code is currently being reviewed but will be release as a 2.0.0 tag.

The main changes consist of changing the constructor of the samplePDFFD object to take a lot of the information from the sample yaml file. This reduces the amount of code that need to be written in each samplePDFDune object.

We now also use the NuOscillator object to calculate oscillation probabilities in the core code.

The samplePDF objects in DUNE have also been slightly renamed to no longer include the word "base" in them as they are not intended to be base classes.

Lots of files have also been removed to reduce the clutter. A few additional exes will be added shortly for people to use as clear examples. There is also a "factory" method to initialise the different samplePDF instances meaning that a lot of executables can now be shared across the beam and atmoshperic samples.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,18 +1,13 @@
 set(CMAKE_VERBOSE_MAKEFILE ON)
 
 # CMake version check
-cmake_minimum_required(VERSION 3.17 FATAL_ERROR)  
+cmake_minimum_required(VERSION 3.14 FATAL_ERROR)
 
-SET(MaCh3_DUNE_VERSION 0.0.1)
+SET(MaCh3DUNE_VERSION 0.0.0)
```
Comment (dbarrow257) [inline CMakeLists.txt:6]: Should increment the version number

File: CMakeLists.txt [hunk 2]
```diff
@@ -29,38 +24,172 @@ find_program(CMAKE_CXX_COMPILER NAMES $ENV{CXX} g++ PATHS ENV PATH NO_DEFAULT_PA
...
 
-#Go get the core MAch3, this will be public one day....
-#This is over-riden with -DCPM_MaCh3_SOURCE=/path/to/MaCh3/folder
 CPMFindPackage(
-    NAME MaCh3
-	GIT_TAG "DUNECore2024"
-    GITHUB_REPOSITORY mach3-software/MaCh3
-)
+      NAME CMakeModules
+      GIT_TAG v0.0.11
```
Comment (dbarrow257) [inline CMakeLists.txt:36]: We should consider incrementing this version

File: CMakeLists.txt [hunk 3]
```diff
@@ -29,38 +24,172 @@ find_program(CMAKE_CXX_COMPILER NAMES $ENV{CXX} g++ PATHS ENV PATH NO_DEFAULT_PA
...
-#Compile and link options (Should be in experiment project but copied in case CORE is built alone) 
-add_compile_options(-g -O3 -Wextra -fopenmp)
-add_link_options(-fopenmp)
+find_package(duneanaobj)
 
+if(NOT duneanaobj_FOUND)
+  CPMFindPackage(
+    NAME duneanaobj
+    GIT_TAG "patch-1"
+    GITHUB_REPOSITORY dbarrow257/duneanaobj
```
Comment (dbarrow257) [inline CMakeLists.txt:60]: We need to do something better with this... but it's ok for now

File: CMakeLists.txt [hunk 4]
```diff
@@ -29,38 +24,172 @@ find_program(CMAKE_CXX_COMPILER NAMES $ENV{CXX} g++ PATHS ENV PATH NO_DEFAULT_PA
...
-if(${CPU_ONLY})
-  add_compile_options(-DCPU_ONLY)
+##################################  MaCh3  ######################################
+#If MaCh3 was sourced find it, otherwise use CPM
+find_package(MaCh3)
+
+if(NOT MaCh3_FOUND)
+  CPMFindPackage(
+      NAME MaCh3
+      GIT_TAG "feature_OscClassMagic"
```
Comment (dbarrow257) [inline CMakeLists.txt:77]: Will need to be changed once core is merged

File: configs/CovObjs/MaCh3Modes.yaml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#########################################
+Title: "DUNE Interaction modes - Still to Implement"
```
Comment (dbarrow257) [inline configs/CovObjs/MaCh3Modes.yaml:2]: Obviously this still needs completing - Should add as an issue

File: configs/CovObjs/OscCov_PDG2021_v2.yaml [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+      ParameterName: baseline
+
+
+    DetID: 984
+    Error: 1.0
+    FlatPrior: true
+    ParameterBounds: [0, 999]
+    ParameterGroup: Osc
+    ParameterValues:
+      Generated: 295
```
Comment (dbarrow257) [inline configs/CovObjs/OscCov_PDG2021_v2.yaml:134]: Should set DUNE's actual baseline

File: configs/CovObjs/xsec_covariance_DUNE_systs_2022a_FD_v3.yaml [hunk 7]
```diff
@@ -0,0 +1,22 @@
+Systematics:
```
Comment (dbarrow257) [inline configs/CovObjs/xsec_covariance_DUNE_systs_2022a_FD_v3.yaml:1]: Do we need this? Seems like a dummy used yaml?

File: configs/Samples/AtmSample_nueselec.yaml [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+  XVarStr: "TrueNeutrinoEnergy"
+  XVarBins: [0., 0.25, 0.5, 0.75, 1.,  1.25, 1.5, 1.75, 2., 2.25, 2.5, 2.75, 3., 3.25, 3.5, 3.75, 4., 4.25, 4.5, 4.75, 5, 5.5, 6, 6.5, 7, 7.5, 8, 8.5, 9, 9.5, 10, 1e8]
+SampleBools:
+  IsELike: yes
+InputFiles:
+  mtupleprefix: "inputs/Atmospherics/CAFs/atm_hd"
+  mtuplesuffix: "_nueselec.root"
+  splineprefix: ""
+  splinesuffix: ""
+DetID: 24
```
Comment (dbarrow257) [inline configs/Samples/AtmSample_nueselec.yaml:20]: Right now, I think the DetID treatment in DUNE is not really used to distinguish samples - should add as issue

File: samplePDFDUNE/samplePDFDUNEAtm.cpp [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+  for (int i=0;i<nSamples;i++) {
+    struct dunemc_base obj = dunemc_base();
+    dunemcSamples.push_back(obj);
+  }
+  
+  IsELike = SampleManager->raw()["SampleBools"]["IsELike"].as<bool>();
+}
+
+void samplePDFDUNEAtm::SetupSplines() {
+  splineFile = nullptr;
```
Comment (dbarrow257) [inline samplePDFDUNE/samplePDFDUNEAtm.cpp:26]: Add as an issue that currently atmospherics can't deal with splines

File: samplePDFDUNE/samplePDFDUNEAtm.cpp [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+  gErrorIgnoreLevel = CurrErrorLevel;
+  
+  return duneobj->nEvents;
+}
+
+void samplePDFDUNEAtm::setupFDMC(int iSample) {
+  dunemc_base *duneobj = &(dunemcSamples[iSample]);
+  fdmc_base *fdobj = &(MCSamples[iSample]);  
+
+  //a bit of a hack, make sure that this is only set if you're an atmoshperic object
```
Comment (dbarrow257) [inline samplePDFDUNE/samplePDFDUNEAtm.cpp:124]: Not sure this is a "hack"

File: samplePDFDUNE/samplePDFDUNEAtm.cpp [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+  
+  return KinematicValue;
+}
+
+double* samplePDFDUNEAtm::ReturnKinematicParameterByReference(double KinematicVariable, int iSample, int iEvent) {
+  KinematicTypes KinPar = (KinematicTypes) std::round(KinematicVariable);
+  return ReturnKinematicParameterByReference(KinPar,iSample,iEvent);
+}
+
+double* samplePDFDUNEAtm::ReturnKinematicParameterByReference(std::string KinematicParameter, int iSample, int iEvent) {
```
Comment (dbarrow257) [inline samplePDFDUNE/samplePDFDUNEAtm.cpp:172]: Should move a lot of this to core, but need to tie into Luke's approach

File: samplePDFDUNE/samplePDFDUNEAtm.h [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+ protected:
+  void Init();
+  int setupExperimentMC(int iSample);
+  void setupFDMC(int iSample);
+
+  void SetupWeightPointers();
+  void SetupSplines();
+  
+  //DB functions which could be initialised to do something which is non-trivial
+  double CalcXsecWeightFunc(int iSample, int iEvent) {return 1.;}
```
Comment (dbarrow257) [inline samplePDFDUNE/samplePDFDUNEAtm.h:28]: Do we need both of these functions?

File: scripts/link_files.sh [hunk 13]
```diff
@@ -1,15 +1,11 @@
 #!/bin/bash
 #
 # A script to link required files to the proper place i.e. where the sample config files will look for them
-# If running on IC machines this will work out of the box
-# If not change FILESDIR to point to where the files live
 
 MACH3DIR=`pwd`
+#FILESDIR=/vols/t2k/users/ljw20/data/DUNE_2021/DUNE_2021_splines_tdr_v8
```
Comment (dbarrow257) [inline scripts/link_files.sh:6]: This file seems incredibly hardcoded to be used at Imperial - can delete?

File: setup.sh [hunk 14]
```diff
@@ -1,29 +1,11 @@
...
-  source /home/stfc/eisext13/root/bin/thisroot.sh
-  export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/home/stfc/eisext13/root/lib
-  export PATH=$PATH:/home/stfc/eisext13/root/bin
-  export ROOTLIBDIR=/home/stfc/eisext13/root/lib
-  export ROOTINCLUDEDIR=/home/stfc/eisext13/root/include
-  module load cuda/8.0.44
-  export CUDAPATH=${CUDA_HOME}
-fi
+source ./build/bin/setup.MaCh3DUNE.sh
+source ../MaCh3_core_dev/build/bin/setup.MaCh3.sh
```
Comment (dbarrow257) [inline setup.sh:4]: Is any of this actually needed? Seems hardcoded to someones (presumably Ed's?) setup with hardcoded relative paths

File: src/Fit.cpp [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+  //Create samplePDFSKBase Objs
+  std::cout << "Loading T2K samples.." << "\n" << std::endl;
+  std::vector<samplePDFFDBase*> DUNEPdfs;
+  MakeMaCh3DuneInstance(FitManager, DUNEPdfs, xsec, osc); 
+  //Setup the cross-section parameters
+  //This should get the prior values.
+  std::vector<double> XsecParVals = xsec->getNominalArray();
+
+  xsec->setParameters(XsecParVals);
+  xsec->setStepScale(FitManager->raw()["General"]["Systematics"]["XsecStepScale"].as<double>());
```
Comment (dbarrow257) [inline src/Fit.cpp:42]: Shouldn't this be in the factory?

File: src/Fit.cpp [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+  //###########################################################################################################
+  // Set covariance objects equal to output of previous chain
+  
+  std::vector<double> oscparstarts;
+  std::map<TString,std::vector<double> > parstarts;
+  int lastStep = 0;
+
+  bool StartFromPreviousChain = GetFromManager(FitManager->raw()["General"]["StartFromPos"], false);
+
+  if(StartFromPreviousChain) {//Start from values at the end of an already run chain
```
Comment (dbarrow257) [inline src/Fit.cpp:81]: I thought this could be done in core code now - if not, we should add an issue so that it could be

**Review by dbarrow257 — COMMENTED** [review]
Minor comments left to address

File: samplePDFDUNE/samplePDFDUNEBeamND.cpp [hunk 17]
```diff
@@ -0,0 +1,467 @@
...
+      MCSamples[i].total_weight_pointers[j][2] = MCSamples[i].osc_w_pointer[j];
+      MCSamples[i].total_weight_pointers[j][3] = &(dunendmcSamples[i].rw_berpaacvwgt[j]);
+      MCSamples[i].total_weight_pointers[j][4] = &(dunendmcSamples[i].flux_w[j]);
+      MCSamples[i].total_weight_pointers[j][5] = &(MCSamples[i].xsec_w[j]);
+    }
+  }
+}
+
+int samplePDFDUNEBeamND::setupExperimentMC(int iSample) {
+  const char *sampleFile = (mtupleprefix+mtuple_files[iSample]+mtuplesuffix).c_str();
```
Comment (luketpickering) [inline samplePDFDUNE/samplePDFDUNEBeamND.cpp:172]: this is wrong, it should be 

```char const *sampleFile = mtuple_files[iSample].c_str()```

looking at how the config is parsed in base and FD/Atm DUNE samplePDFs

File: samplePDFDUNE/samplePDFDUNEBeamNDGar.cpp [hunk 18]
```diff
@@ -0,0 +1,472 @@
...
+      MCSamples[i].total_weight_pointers[j][2] = MCSamples[i].osc_w_pointer[j];
+      MCSamples[i].total_weight_pointers[j][3] = &(dunendgarmcSamples[i].rw_berpaacvwgt[j]);
+      MCSamples[i].total_weight_pointers[j][4] = &(dunendgarmcSamples[i].flux_w[j]);
+      MCSamples[i].total_weight_pointers[j][5] = &(MCSamples[i].xsec_w[j]);
+    }
+  }
+}
+
+int samplePDFDUNEBeamNDGar::setupExperimentMC(int iSample) {
+  const char *sampleFile = (mtupleprefix+mtuple_files[iSample]+mtuplesuffix).c_str();
```
Comment (luketpickering) [inline samplePDFDUNE/samplePDFDUNEBeamNDGar.cpp:62]: also wrong here, as in ND code

File: samplePDFDUNE/samplePDFDUNEBeamFD.cpp [hunk 19]
```diff
@@ -0,0 +1,839 @@
...
+      MCSamples[i].total_weight_pointers[j][4] = &(dunemcSamples[i].flux_w[j]);
+      MCSamples[i].total_weight_pointers[j][5] = &(MCSamples[i].xsec_w[j]);
+    }
+  }
+}
+
+
+int samplePDFDUNEBeamFD::setupExperimentMC(int iSample) {
+  const char *sampleFile = (mtuple_files[iSample]).c_str();
+  dunemc_base *duneobj = &(dunemcSamples[iSample]);
```
Comment (luketpickering) [inline samplePDFDUNE/samplePDFDUNEBeamFD.cpp:204]: its fairly minor, but this is just unnecessary. We really can do without any bare pointers in most of this code:

```c++
auto &duneobj = dunemcSamples[iSample];

//...
duneobj.bla[i] = _bla;
```

File: samplePDFDUNE/samplePDFDUNEAtm.cpp [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+  Initialise();
+}
+
+samplePDFDUNEAtm::~samplePDFDUNEAtm() {
+}
+
+void samplePDFDUNEAtm::Init() {
+  for (int i=0;i<nSamples;i++) {
+    struct dunemc_base obj = dunemc_base();
+    dunemcSamples.push_back(obj);
```
Comment (luketpickering) [inline samplePDFDUNE/samplePDFDUNEAtm.cpp:19]: can just be:

```c++
dunemcSamples.resize(nSamples);
```

File: samplePDFDUNE/samplePDFDUNEBeamND.cpp (hunk 17 above)
Comment (dbarrow257) [reply samplePDFDUNE/samplePDFDUNEBeamND.cpp]: Fixed - you are correct

File: samplePDFDUNE/samplePDFDUNEBeamNDGar.cpp (hunk 18 above)
Comment (dbarrow257) [reply samplePDFDUNE/samplePDFDUNEBeamNDGar.cpp]: Fixed, you are correct

File: samplePDFDUNE/samplePDFDUNEAtm.cpp (hunk 20 above)
Comment (dbarrow257) [reply samplePDFDUNE/samplePDFDUNEAtm.cpp]: Done

File: samplePDFDUNE/samplePDFDUNEAtm.h (hunk 12 above)
Comment (dbarrow257) [reply samplePDFDUNE/samplePDFDUNEAtm.h:28]: Yes, it looks like they do two different tasks. One being a weight, and the other being a shift.

File: src/Fit.cpp (hunk 15 above)
Comment (dbarrow257) [reply src/Fit.cpp]: Done

**Review by dbarrow257 — COMMENTED** [review]
Only minor comments left which modify code style but don't change behaviour

File: setup.sh (hunk 14 above)
Comment (dbarrow257) [reply setup.sh]: Removed

File: CMakeLists.txt (hunk 4 above)
Comment (dbarrow257) [reply CMakeLists.txt]: Now pulls develop

File: scripts/link_files.sh (hunk 13 above)
Comment (dbarrow257) [reply scripts/link_files.sh:6]: Left for now - should think about just directly putting this information into the README

**Review by luketpickering — APPROVED** [review]

**Pull request merged**